File: backend/app/models/farm.ts
import { DateTime } from 'luxon'
import { BaseModel, column, belongsTo } from '@adonisjs/lucid/orm'
import type { BelongsTo } from '@adonisjs/lucid/types/relations'
import User from './user.js'

export default class Farm extends BaseModel {
  @column({ isPrimary: true })
  declare id: number

  @column()
  declare userId: number

  @column()
  declare name: string

  @column()
  declare type: 'GRAPES' | 'WHEAT' | 'CORN' | 'TOMATOES' | 'OLIVES' | 'DATES'

  @column({
    prepare: (value: Array<{ latitude: number; longitude: number }>) => JSON.stringify(value),
    consume: (value: string) => JSON.parse(value),
  })
  declare polygon: Array<{ latitude: number; longitude: number }>

  @column()
  declare area: number

  @belongsTo(() => User)
  declare user: BelongsTo<typeof User>

  @column.dateTime({ autoCreate: true })
  declare createdAt: DateTime

  @column.dateTime({ autoCreate: true, autoUpdate: true })
  declare updatedAt: DateTime | null
}

File: app/lib/features/farms/domain/entities/farm.dart
enum FarmType {
  grapes,
  wheat,
  corn,
  tomatoes,
  olives,
  dates,
}

class LatLng {
  final double latitude;
  final double longitude;

  const LatLng({
    required this.latitude,
    required this.longitude,
  });

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is LatLng &&
        other.latitude == latitude &&
        other.longitude == longitude;
  }

  @override
  int get hashCode => Object.hash(latitude, longitude);
}

class Farm {
  final String id;
  final String userId;
  final String name;
  final FarmType type;
  final List<LatLng> polygon;
  final double area;

  const Farm({
    required this.id,
    required this.userId,
    required this.name,
    required this.type,
    required this.polygon,
    required this.area,
  });

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is Farm &&
        other.id == id &&
        other.userId == userId &&
        other.name == name &&
        other.type == type &&
        other.area == area;
  }

  @override
  int get hashCode => Object.hash(id, userId, name, type, area);
}

